src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,293 @@
+        }
+        if header.header != HDRS {
+            let i = header.header;
+            let p = unsafe { &header.header };
+            write!(w, "header.header is {:x}, not {:x} @ {:x}\n\r\n", i, HDRS, p).unwrap();
+            return Err("magic number missing: header.header != 0x53726448");
+        }
+        // only accept version >= 2.12
+        if header.version < 0x020c {
+            let version = header.version;
+            write!(w, "kernel version too old: 0x{:04x}", version).unwrap();
+            return Err("kernel version too old");
+        }
+        // if header.xloadflags | XLF_KERNEL_64 == 0 {
+        //     return Err("kernel has no 64-bit entry point");
+        // }
+        // if header.xloadflags | XLF_CAN_BE_LOADED_ABOVE_4G == 0 {
+        //     return Err("kernel cannot be loaded above 4GiB");
+        // }
+        // if false && header.relocatable_kernel == 0 {
+        //     return Err("kernel is not relocatable");
+        // }
+
+        // calculate offsets
+        // bootparam offset
+        // can we just stick with 90000 for now?
+        //let bp_offset = self.mem_size - PAGE_SIZE;
+        //let cmd_line_offset = bp_offset - PAGE_SIZE;
+
+        let mut bp = BootParams::default();
+        bp.hdr = header;
+
+        // load kernel
+        let mut kernel_offset = (bp.hdr.setup_sects as usize + 1) * 512;
+
+        write!(w, "Kernel offset is {:x}\r\n", kernel_offset).unwrap();
+        // Copy from driver into segment.
+        let mut buf = [0u8; 512];
+        let mut amt: usize = 0;
+        writeln!(w, "Copy {:x} to {:x} for {:x}", kernel_offset + self.rom_base, self.load, 0x300000).unwrap();
+        let mut load = self.load;
+        loop {
+            if amt > 0x300000 {
+                break;
+            }
+            let size = match rom.pread(&mut buf, kernel_offset) {
+                Ok(x) => x,
+                _x => break,
+            };
+            //write!(w, "Copy {:x} to {:x} for {:x}\n", buf.as_ptr() as u64, self.load, size).unwrap();
+            unsafe { copy(buf.as_ptr(), load as *mut u8, size) };
+            kernel_offset += size;
+            load += size;
+            amt += size;
+        }
+
+        // setup e820 tables
+
+        // The first page is always reserved.
+        let entry_first_page = E820Entry { addr: 0, size: PAGE_SIZE as u64, r#type: E820_RESERVED };
+        // a tiny bit of low memory for trampoline
+        let entry_low = E820Entry { addr: PAGE_SIZE as u64, size: (LOW_MEM_64K - PAGE_SIZE) as u64, r#type: E820_RAM };
+        // memory from 64K to LOW_MEM_1M is reserved
+        let entry_reserved = E820Entry { addr: LOW_MEM_64K as u64, size: (LOW_MEM_1M - LOW_MEM_64K) as u64, r#type: E820_RESERVED };
+        // LOW_MEM_1M to low_mem_size for ramdisk and multiboot
+        let entry_low_main = E820Entry { addr: LOW_MEM_1M as u64, size: (self.low_mem_size - LOW_MEM_1M as u64), r#type: E820_RAM };
+        // main memory above 4GB
```

Comment:
What about the region between low_mem_size to 4G? why is there a gap?

Reply:
what values are you talking? You do need some gap for pci address space.

I think I set it to be a 2G gap.

Reply:
Leave a comment to say so?